### PR TITLE
simple F# command line parser

### PR DESCRIPTION
**Pull request by brettfo**
This enables simple F# command line parsing.  There are only 4 different things that can be parsed which makes this parser very small:

1.  Metadata references.  For fsc.exe these all start with `-r:` or `--reference:`.
2.  Analyzer references.  F# doesn't support these so it's always an empty collection.
3.  Source files.  Non-option arguments (e.g., no leading `-`) that end in `.fs`, `.fsi`, etc.
4.  Additional files.  F# doesn't support these so it's always an empty collection.

This required updating the Roslyn references to bring in the newly exposed constructors in `CommandLineSourceFile` and `CommandLineReference`.

@KevinRansom @davkean @srivatsn 

Edit: I've verified these changes locally along with dotnet/roslyn#19566.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs [hunk 1]
```diff
@@ -2,19 +2,56 @@
...
+                        case ".fs":
+                        case ".fsi":
+                        case ".fsx":
+                        case ".fsscript":
+                        case ".ml":
+                        case ".mli":
+                            sourceFiles.Add(new CommandLineSourceFile(arg, isScript: (extension == ".fsx") || (extension == ".fsscript")));
+                            break;
+                        default:
+                            additionalFiles.Add(new CommandLineSourceFile(arg, isScript: false));
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs]: I don't think this is right - the F# LS doesn't understand additional files yet since there's no analyzers support. In C#\VB, additionalfiles are specific item types. I think it'd be best to make this always empty until real analyzers support is implemented.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs [hunk 2]
```diff
@@ -2,19 +2,56 @@
...
+using System.IO;
+using Microsoft.CodeAnalysis;
 
 namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
 {
     [Export(typeof(IParseBuildOptions))]
     [AppliesTo(ProjectCapability.FSharp)]
     internal class FSharpParseBuildOptions : IParseBuildOptions
     {
+        private const string ReferencePrefix = "-r:";
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs]: Is there no long form like -reference for F#?

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs [hunk 3]
```diff
@@ -2,19 +2,56 @@
...
+                            break;
+                    }
+                }
+            }
+
+            return new BuildOptions(
+                sourceFiles,
+                additionalFiles,
+                metadataReferences,
+                new CommandLineAnalyzerReference[0]);
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs]: Use Array.Empty. @davkean, this will conflict when the LS branch gets merged because it's been moved to immutable arrays.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs (hunk 2 above)
Comment (brettfo) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs]: There is a `--reference` option, but that's never exposed via the design time build; only `-r` is used.  I'll be sure to add `--reference` support to the official parser that will live in the F# code.

**Comment by brettfo** [comment]
Created F# issue Microsoft/visualfsharp#3074 to track the creation of a proper command line parser.

File: build/Targets/References.targets [hunk 4]
```diff
@@ -66,11 +66,11 @@
 
         <!-- NuGet Dependencies -->
         <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
-        <PackageReference Include="System.Reflection.Metadata" Version="1.4.1-beta-24430-01" />
+        <PackageReference Include="System.Reflection.Metadata" Version="1.4.2" />
         <PackageReference Include="System.ValueTuple" Version="4.3.0" />
         <PackageReference Include="Microsoft.Composition" Version="1.0.27" />
         <PackageReference Include="Newtonsoft.Json" Version="8.0.3" />
-        <PackageReference Include="Microsoft.VisualStudio.LanguageServices" Version="2.0.0-rc2" />
+        <PackageReference Include="Microsoft.VisualStudio.LanguageServices" Version="2.3.0-beta1-61703-10" />
```
Comment (srivatsn) [inline build/Targets/References.targets]: This is unrelated to this PR - correct? @jmarolf can you get this in sooner - it's causing package downgrade warnings and failed designtime builds in the IDE.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs (hunk 1 above)
Comment (brettfo) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs]: Good catch.  Fixed.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs (hunk 3 above)
Comment (brettfo) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp/ProjectSystem/LanguageServices/FSharpParseBuildOptions.cs]: Done.  Also, I've pre-emptively changed these to `.ToImmutableArray()` and `ImmutableArray<T>.Empty`.

File: build/Targets/References.targets (hunk 4 above)
Comment (brettfo) [inline build/Targets/References.targets]: No, this is necessary because I did some work in Roslyn to expose the constructors for the `CommandLineSourceFile` and `CommandLineReference` structs and this is the first build that contains these changes.  This won't build otherwise.

**Comment by srivatsn** [comment]
@brettfo can you fill out the template and tag Matt

**Comment by brettfo** [comment]
Tagging @MattGertz

**Customer scenario**

Enables the new project system to properly parse source files and metadata references from the provided command line options.

**Bugs this fixes:** 

No direct fix, but enables future work, including by community members.

**Workarounds, if any**

N/A.

**Risk**

Low.  Only affects F# .NET Core projects and adds missing functionality.

**Performance impact**

Low.  This parser is much simpler (and therefore faster) than the fallback C# parser that was previously specified.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

N/A.

**How was the bug found?**

Part of ongoing F# project system work.

**Comment by dsyme** [comment]
What about ``--define:XYZ``, are they needed? thanks

**Comment by brettfo** [comment]
The only outputs consumed by this parser (for C# and VB, too) are source files, additional files, metadata references, and analyzer references.  The other options are handled by the language service code, but eventually will be migrated to use [`SetOptions()`](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs,d1d4a99a38bdfb75).
